### PR TITLE
Mesh closest point

### LINKED ISSUE #416: ClosestPoint() implementation for meshes
### Discussed in https://github.com/GSharker/G-Shark/discussions/415

<div type='discussions-op-text'>

<sup>Originally posted by **iltabe** May 30, 2023</sup>
Hi guys!

I know some of you for different reasons! Ciao!

For a client I'm building an algorithm that requires some geometrical operations. I've looked into G-Shark and started to use it. It looks promising and does almost everything I need (basic vector, point, plane stuff).
Thanks guys for having give birth to this nice project!

The only thing I miss is a `mesh.ClosestPoint(Point3 pt)` method... Is it in the todo list by any luck?
Thanks!
Gianluca</div>

It would be nice to implement some functionality for meshes, especially `myMesh.ClosestPoint(Point3 test_point)`.
The behaviour I expect is to get back from the function the real closest point:

1. on a mesh vertex, if the mesh vertex is the closest point to the test point
2. on a mesh face boundary (edge) if that is the closest point to the test point
3. a point inside a mesh face if the length of the line perpendicular to the face (and passing per test_point) is shorter to any other line connecting test_point to the mesh.

Long story short: the closest point! :)

### PR DESCRIPTION
**Pull request by Sophielelerre**
### What type of PR is this? (check all applicable)

- [ ] ♻️ Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 👷 Optimization
- [ ] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

### Description

This PR adds a ClosestPoint() implementation for meshes. The method takes a mesh and a test point and returns the closest point to the test point on the mesh. The algorithm finds the closest point to all the faces, then the absolute closest. This point can be either on a mesh vertex, mesh edge, or inside a mesh face. The method can be used with triangle meshes, quad meshes, and Ngon meshes. Any Ngons faces are triangulated using vertices' centroid and consecutive vertices. Quads are treated as Ngons. I did not work on the tests yet, I'd like some feedback first.

### Related Tickets & Documents

Please use this format link issue numbers: Fixes #416
https://github.com/GSharker/G-Shark/issues/416#issue-1736250025

### Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [x] 🙋 no, because I need help

### Added to documentation?

- [ ] 📓 docs
- [x] 🙅 no documentation needed

### CODE REVIEW

File: src/GShark/Geometry/Mesh.cs [hunk 1]
```diff
@@ -547,6 +547,122 @@ private FaceData CountFaceEdges()
             return data;
         }
 
+        /// <summary>
+        /// Checks if point is in or out a triangle.
```
Comment (d3ssy) [inline src/GShark/Geometry/Mesh.cs:551]: inside or outside of a triangle

File: src/GShark/Geometry/Mesh.cs [hunk 2]
```diff
@@ -547,6 +547,122 @@ private FaceData CountFaceEdges()
...
+
+
+        /// <summary>
+        /// Finds the closest point to a triangle.
+        /// </summary>
+        /// <param name="trianglePoints">Triangle vertices.</param>
+        /// <param name="point">Test point.</param>
+        /// <returns>Closest point to the triangle.</returns>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown when trianglePoints is not 3 points.</exception>
+        private Point3 ClosestPointToTriangle(Point3[] trianglePoints, Point3 point)
```
Comment (d3ssy) [inline src/GShark/Geometry/Mesh.cs:595]: - ClosestPoint is fine as method name on Mesh, trianglular polygon implied by input args
- Same comments regarding Polygon input vs point array as above
- Same comments regarding naming

File: src/GShark/Geometry/Mesh.cs [hunk 3]
```diff
@@ -547,6 +547,122 @@ private FaceData CountFaceEdges()
...
+            foreach (Line segment in segments)
+            {
+                Point3 segmentClosestPoint = segment.ClosestPoint(point);
+                edgesClosestPoints.Add(segmentClosestPoint);
+            }
+            return Point3.CloudClosestPoint(edgesClosestPoints, point);
+        }
+
+        /// <summary>
+        /// Finds the closest point to a mesh.
```
Comment (d3ssy) [inline src/GShark/Geometry/Mesh.cs:624]: A bit more detail to differentiate the different ClosestPoint methods would be helpful in the summary docstring

File: src/GShark/Geometry/Mesh.cs [hunk 4]
```diff
@@ -547,6 +547,122 @@ private FaceData CountFaceEdges()
...
+                    Point3 ngonCentre = Point3.AveragePoint(faceVertices);
+                    var TrianglesCP = new List<Point3>();
+
+                    for (int i = 0; i < faceVertices .Count; i++)
+                    {
+                        Point3[] trianglePoints = new Point3[] { ngonCentre, faceVertices[i], faceVertices[(i + 1)% faceVertices.Count] };
+                        Point3 ClosestPointToNgon = ClosestPointToTriangle(trianglePoints, point);
+                        TrianglesCP.Add(ClosestPointToNgon);
+                    }
+                    Point3 ClosestPoint = Point3.CloudClosestPoint(TrianglesCP, point);
```
Comment (d3ssy) [inline src/GShark/Geometry/Mesh.cs:659]: lowercase for vars

File: src/GShark/Geometry/Mesh.cs [hunk 5]
```diff
@@ -547,6 +547,122 @@ private FaceData CountFaceEdges()
...
+                {
+                    var trianglePoints = new Point3[] { faceVertices[0], faceVertices[1], faceVertices[2] };
+                    Point3 closestPoint0 = ClosestPointToTriangle(trianglePoints, point);
+                    closerPointToFaceList.Add(closestPoint0);
+                }
+                else 
+                {
+                    // Ngon triangulation using the vertices' centroid and consecutive vertices to create the triangles 
+                    Point3 ngonCentre = Point3.AveragePoint(faceVertices);
+                    var TrianglesCP = new List<Point3>();
```
Comment (d3ssy) [inline src/GShark/Geometry/Mesh.cs:651]: lowercase for vars

File: src/GShark/Geometry/Mesh.cs [hunk 6]
```diff
@@ -547,6 +547,122 @@ private FaceData CountFaceEdges()
...
+        /// </summary>
+        /// <param name="point">Test point.</param>
+        /// <returns>Mesh Closest Point.</returns>
+        /// <remark>Method might not work with concave Ngon faces.</remark>
+        public Point3 ClosestPoint(Point3 point)
+        {
+            List<MeshVertex> meshVertices = Vertices;
+            // All faces check 
+            IEnumerable<MeshFace> allFaces = this.Faces;
+            var closerPointToFaceList = new List<Point3>();
```
Comment (d3ssy) [inline src/GShark/Geometry/Mesh.cs:634]: closer or closest?

File: src/GShark/Geometry/Mesh.cs [hunk 7]
```diff
@@ -547,6 +547,122 @@ private FaceData CountFaceEdges()
...
+                else 
+                {
+                    // Ngon triangulation using the vertices' centroid and consecutive vertices to create the triangles 
+                    Point3 ngonCentre = Point3.AveragePoint(faceVertices);
+                    var TrianglesCP = new List<Point3>();
+
+                    for (int i = 0; i < faceVertices .Count; i++)
+                    {
+                        Point3[] trianglePoints = new Point3[] { ngonCentre, faceVertices[i], faceVertices[(i + 1)% faceVertices.Count] };
+                        Point3 ClosestPointToNgon = ClosestPointToTriangle(trianglePoints, point);
```
Comment (d3ssy) [inline src/GShark/Geometry/Mesh.cs:656]: lowercase for vars

File: src/GShark/Geometry/Point3.cs [hunk 8]
```diff
@@ -717,7 +717,46 @@ public int InPolygon(Polygon polygon)
...
             return inside ? 1 : -1;
         }
-        
+
+        /// <summary>
+        /// Finds average point from a cloud of points.  
+        /// </summary>
+        /// <param name="points"></param>
+        /// <returns>The average point.</returns>
+        public static Point3 AveragePoint (IEnumerable<Point3> points)
```
Comment (d3ssy) [inline src/GShark/Geometry/Point3.cs:726]: Already implemented as Centroid() method in Point3 
https://github.com/GSharker/G-Shark/blob/76819e0116c12387fef8d86ffbecf9d608e9f705/src/GShark/Geometry/Point3.cs#L520

File: src/GShark/Geometry/Point3.cs [hunk 9]
```diff
@@ -717,7 +717,46 @@ public int InPolygon(Polygon polygon)
...
+            foreach (Point3 point in points)
+            {
+                pointTotSum += point;
+            }
+            Point3 average = pointTotSum / points.Count();
+            return average;
+        }
+
+        /// <summary>
+        /// Finds closest point to a test point from a cloud of points.
```
Comment (d3ssy) [inline src/GShark/Geometry/Point3.cs:738]: This should be moved to Point3 class as `Point3.ClosestPoint(Point3[] points, Point3 testPoint)`. I don't think the "cloud" terminology is useful here as we do not have a specialized PointCloud class.

**Review by d3ssy — CHANGES_REQUESTED** [review]
Thank you for this contribution! Please check comments and update code accordingly. Mostly cosmetic and readibility comments.

Please also add the relevant unit tests in the appropriate test classes for the new methods. For closest point methods edge and point coincidence edge cases will be important to include.

There is also another PR open for PointInPolygon which may or may not be useful. https://github.com/GSharker/G-Shark/pull/419
